File: src/dsc/runtime/evaluator.py
"""Deterministic condition evaluator.

Evaluates ConditionExpr AST nodes against observation dictionaries.
This is the hot path at runtime — pure functions, no side effects.
"""

from __future__ import annotations

import re
from typing import Any

from dsc.models.conditions import (
    AlwaysTrue,
    ConditionExpr,
    ConditionGroup,
    FieldCondition,
    LogicOperator,
    Operator,
)

# Sentinel for missing fields
_MISSING = object()


def resolve_field(observation: dict, field_path: str) -> Any:
    """Resolve a dot-separated field path in an observation dict.

    Returns _MISSING if the field doesn't exist at any level.

    Examples:
        resolve_field({"user": {"intent": "refund"}}, "user.intent") → "refund"
        resolve_field({"x": 1}, "y") → _MISSING
    """
    current = observation
    for part in field_path.split("."):
        if isinstance(current, dict) and part in current:
            current = current[part]
        else:
            return _MISSING
    return current


def evaluate_field_condition(condition: FieldCondition, observation: dict) -> bool:
    """Evaluate a single field condition against an observation."""
    value = resolve_field(observation, condition.field)
    if value is _MISSING:
        return False

    expected = condition.value
    op = condition.operator

    try:
        if op == Operator.EQ:
            return value == expected
        elif op == Operator.NE:
            return value != expected
        elif op == Operator.GT:
            return value > expected
        elif op == Operator.LT:
            return value < expected
        elif op == Operator.GTE:
            return value >= expected
        elif op == Operator.LTE:
            return value <= expected
        elif op == Operator.IN:
            return value in expected
        elif op == Operator.NOT_IN:
            return value not in expected
        elif op == Operator.CONTAINS:
            return expected in value
        elif op == Operator.MATCHES:
            return bool(re.search(expected, str(value)))
    except (TypeError, ValueError):
        return False

    return False


def evaluate(condition: ConditionExpr, observation: dict) -> bool:
    """Evaluate any condition expression against an observation.

    This is the main entry point for condition evaluation.
    """
    if isinstance(condition, AlwaysTrue):
        return True

    if isinstance(condition, FieldCondition):
        return evaluate_field_condition(condition, observation)

    if isinstance(condition, ConditionGroup):
        return _evaluate_group(condition, observation)

    raise TypeError(f"Unknown condition type: {type(condition)}")


def _evaluate_group(group: ConditionGroup, observation: dict) -> bool:
    """Evaluate a compound condition group."""
    if group.logic == LogicOperator.AND:
        return all(evaluate(c, observation) for c in group.conditions)
    elif group.logic == LogicOperator.OR:
        return any(evaluate(c, observation) for c in group.conditions)
    elif group.logic == LogicOperator.NOT:
        if len(group.conditions) != 1:
            raise ValueError("NOT condition must have exactly one sub-condition")
        return not evaluate(group.conditions[0], observation)
    raise ValueError(f"Unknown logic operator: {group.logic}")

File: openclaw/plugin/src/evaluator.ts
/**
 * DSC Condition Evaluator — TypeScript port of the Python runtime evaluator.
 *
 * Pure function: evaluate(condition, observation) -> boolean
 * Handles nested field paths, type mismatches, and all 10 operators.
 */

export interface FieldCondition {
  type: "field";
  field: string;
  operator: string;
  value: unknown;
}

export interface ConditionGroup {
  type: "group";
  logic: "and" | "or" | "not";
  conditions: Condition[];
}

export interface AlwaysTrue {
  type: "always_true";
}

export type Condition = FieldCondition | ConditionGroup | AlwaysTrue;

export interface Transition {
  from_state: string;
  condition: Condition;
  action: string;
  action_params: Record<string, unknown>;
  to_state: string;
  priority: number;
}

export interface CompiledArtifact {
  version: number;
  scenario_id: string;
  scenario_name: string;
  graph: {
    states: Record<string, { description: string; is_terminal: boolean }>;
    transitions: Transition[];
    initial_state: string;
  };
}

/**
 * Resolve a dot-path field from an observation object.
 * e.g., "customer.tier" from { customer: { tier: "vip" } } -> "vip"
 */
function resolveField(path: string, observation: Record<string, unknown>): unknown {
  const parts = path.split(".");
  let current: unknown = observation;

  for (const part of parts) {
    if (current === null || current === undefined || typeof current !== "object") {
      return undefined;
    }
    current = (current as Record<string, unknown>)[part];
  }

  return current;
}

/**
 * Evaluate a single condition against an observation.
 */
export function evaluate(condition: Condition, observation: Record<string, unknown>): boolean {
  switch (condition.type) {
    case "always_true":
      return true;

    case "field":
      return evaluateField(condition, observation);

    case "group":
      return evaluateGroup(condition, observation);

    default:
      return false;
  }
}

function evaluateField(condition: FieldCondition, observation: Record<string, unknown>): boolean {
  const fieldValue = resolveField(condition.field, observation);

  if (fieldValue === undefined) {
    return false;
  }

  const expected = condition.value;

  switch (condition.operator) {
    case "eq":
      return fieldValue === expected;
    case "ne":
      return fieldValue !== expected;
    case "gt":
      return typeof fieldValue === "number" && typeof expected === "number" && fieldValue > expected;
    case "lt":
      return typeof fieldValue === "number" && typeof expected === "number" && fieldValue < expected;
    case "gte":
      return typeof fieldValue === "number" && typeof expected === "number" && fieldValue >= expected;
    case "lte":
      return typeof fieldValue === "number" && typeof expected === "number" && fieldValue <= expected;
    case "in":
      return Array.isArray(expected) && expected.includes(fieldValue);
    case "not_in":
      return Array.isArray(expected) && !expected.includes(fieldValue);
    case "contains":
      return typeof fieldValue === "string" && typeof expected === "string" && fieldValue.includes(expected);
    case "matches":
      if (typeof fieldValue === "string" && typeof expected === "string") {
        try {
          return new RegExp(expected).test(fieldValue);
        } catch {
          return false;
        }
      }
      return false;
    default:
      return false;
  }
}

function evaluateGroup(condition: ConditionGroup, observation: Record<string, unknown>): boolean {
  switch (condition.logic) {
    case "and":
      return condition.conditions.every((c) => evaluate(c, observation));
    case "or":
      return condition.conditions.some((c) => evaluate(c, observation));
    case "not":
      return condition.conditions.length > 0 && !evaluate(condition.conditions[0], observation);
    default:
      return false;
  }
}
